### PR TITLE
DSND-1536 Common code branch (Adds new service / validator classes)

### PR DESCRIPTION
**Pull request by kinpang-CH**
- Implement new ApiService to call resource-changed endpoint
- Adds new Validator classes to validate on acceptable company statuses
- Change Interceptors to use up to date authentication / authorisation

Unblocks [DSND-1536](https://companieshouse.atlassian.net/browse/DSND-1536)

[DSND-1536]: https://companieshouse.atlassian.net/browse/DSND-1536?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by sthompsonCH — APPROVED** [review]

**Review by cgregoryCH — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/company_appointments/interceptor/AuthenticationInterceptor.java [hunk 1]
```diff
@@ -17,28 +18,45 @@ public class AuthenticationInterceptor implements HandlerInterceptor {
...
 
-    private boolean isInvalidIdentityType(HttpServletRequest request) {
-        String identityType = request.getHeader(ERIC_IDENTITY_TYPE);
-        return !("key".equalsIgnoreCase(identityType) || "oauth2".equalsIgnoreCase(identityType));
+    private boolean isKeyAuthorised(HttpServletRequest request, String ericIdentityType) {
+        String[] privileges = authenticationHelper.getApiKeyPrivileges(request);
+
+        return HttpMethod.GET.matches(request.getMethod())
+                || ("Key".equalsIgnoreCase(ericIdentityType)
+                && ArrayUtils.contains(privileges, "internal-app"));
```
Comment (johncook-ch) [inline src/main/java/uk/gov/companieshouse/company_appointments/interceptor/AuthenticationInterceptor.java:56]: Should this be a case-insensitive comparison?

File: src/test/java/uk/gov/companieshouse/company_appointments/ResourceChangedApiServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        verify(apiClientService, times(1)).getInternalApiClient();
+        verify(internalApiClient, times(1)).privateChangedResourceHandler();
+        verify(privateChangedResourceHandler, times(1)).postChangedResource("/resource-changed", changedResource);
+        verify(changedResourcePost, times(1)).execute();
+    }
+
+    private static Stream<Arguments> invokeChsKafkaApiExceptionFixtures() {
+        return Stream.of(
+                Arguments.of(
+                        Named.of("Throws service unavailable exception when response code is HTTP 500",
```
Comment (johncook-ch) [inline src/test/java/uk/gov/companieshouse/company_appointments/ResourceChangedApiServiceTest.java:116]: The name says 500, but the exception uses a code of 503. 
See the next element where the numbers are swapped.
Should the text and codes agree?

**Review by johncook-ch — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/company_appointments/interceptor/AuthenticationInterceptor.java (hunk 1 above)
Comment (kinpang-CH) [reply src/main/java/uk/gov/companieshouse/company_appointments/interceptor/AuthenticationInterceptor.java:56]: Addressed in latest commit

File: src/test/java/uk/gov/companieshouse/company_appointments/ResourceChangedApiServiceTest.java (hunk 2 above)
Comment (kinpang-CH) [reply src/test/java/uk/gov/companieshouse/company_appointments/ResourceChangedApiServiceTest.java:116]: Addressed in latest commit

**Review by johncook-ch — APPROVED** [review]

**Pull request merged**